### PR TITLE
Add partial retake option

### PR DESCRIPTION
**Pull request by kjuej**

### CODE REVIEW

File: src/teaching-element/Assessment/DragDrop.vue [hunk 1]
```diff
@@ -126,6 +128,25 @@ export default {
...
+            this.answersCollection.push(answer);
+          }
+        });
+        this.removeFromGroup(wrongAnswers, groupId);
+      });
+    },
+    removeFromGroup(wrongAnswers, groupId) {
+      if (!wrongAnswers.length) return;
+      wrongAnswers.forEach(answer => {
+        const index = this.userAnswer[groupId].map(e => e.id).indexOf(answer);
```
Comment (abasic) [inline src/teaching-element/Assessment/DragDrop.vue]: There is a [`findIndex`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex) method on an array.

File: src/teaching-element/Assessment/DragDrop.vue [hunk 2]
```diff
@@ -148,8 +169,11 @@ export default {
   watch: {
     retake(val) {
       if (!val) return;
-      this.answersCollection = formatAnswers(this.answers);
-      this.userAnswer = mapValues(this.groups, () => []);
+      if (this.options.partialRetake) this.retakeIncorrect();
```
Comment (abasic) [inline src/teaching-element/Assessment/DragDrop.vue]: Please do an early return.

File: src/teaching-element/Assessment/MatchingQuestion.vue [hunk 3]
```diff
@@ -16,7 +16,7 @@
             v-for="({ dragged, value }, index) in source"
             :key="index"
             class="drag-spot">
-            <span :class="{ dragged }" class="item">{{ value }}</span>
+            <span v-if="!dragged" :class="{ dragged }" class="item">{{ value }}</span>
```
Comment (abasic) [inline src/teaching-element/Assessment/MatchingQuestion.vue]: If `dragged` is `false` then it is shown but the class is not applied. If `dragged` is `true` it is not shown so the element does not exist. So there is no reason to ever apply the dragged `class`.

File: src/teaching-element/Assessment/MatchingQuestion.vue [hunk 4]
```diff
@@ -117,6 +118,17 @@ export default {
     draggable(item) {
       return get(item.answers, '0.value', '');
     },
+    retakeIncorrect() {
+      this.target.forEach(response => {
+        const key = response.answers[0].key;
+        if (this.correct[key] !== response.key) {
+          this.source.map(it => {
```
Comment (abasic) [inline src/teaching-element/Assessment/MatchingQuestion.vue]: `map` returns a new array which is not used here so there is no need to use it here. `forEach` should be used instead.

File: src/teaching-element/Assessment/MatchingQuestion.vue [hunk 5]
```diff
@@ -146,8 +158,11 @@ export default {
   watch: {
     retake(val) {
       if (!val) return;
-      this.initialize();
-      this.update();
+      if (this.options.partialRetake) this.retakeIncorrect();
```
Comment (abasic) [inline src/teaching-element/Assessment/MatchingQuestion.vue]: Use an early return.

File: src/teaching-element/Assessment/DragDrop.vue [hunk 6]
```diff
@@ -126,6 +128,25 @@ export default {
       this.userAnswer[id] = this.userAnswer[id].filter(a => a.id !== answer.id);
       this.answersCollection.push(answer);
     },
+    retakeIncorrect() {
+      Object.keys(this.userAnswer).forEach(groupId => {
+        let wrongAnswers = [];
+        this.userAnswer[groupId].forEach((answer) => {
```
Comment (abasic) [inline src/teaching-element/Assessment/DragDrop.vue]: No need for brackets around `answer`.

File: src/teaching-element/Assessment/DragDrop.vue [hunk 7]
```diff
@@ -126,6 +128,25 @@ export default {
...
+            wrongAnswers.push(answer.id);
+            this.answersCollection.push(answer);
+          }
+        });
+        this.removeFromGroup(wrongAnswers, groupId);
+      });
+    },
+    removeFromGroup(wrongAnswers, groupId) {
+      if (!wrongAnswers.length) return;
+      wrongAnswers.forEach(answer => {
```
Comment (abasic) [inline src/teaching-element/Assessment/DragDrop.vue]: This should be named `answerId` instead.

File: src/teaching-element/Assessment/DragDrop.vue [hunk 8]
```diff
@@ -126,6 +128,25 @@ export default {
       this.userAnswer[id] = this.userAnswer[id].filter(a => a.id !== answer.id);
       this.answersCollection.push(answer);
     },
+    retakeIncorrect() {
+      Object.keys(this.userAnswer).forEach(groupId => {
+        let wrongAnswers = [];
```
Comment (abasic) [inline src/teaching-element/Assessment/DragDrop.vue]: This should be named `wrongAnswersId`.

File: src/teaching-element/Assessment/DragDrop.vue [hunk 9]
```diff
@@ -126,6 +128,25 @@ export default {
...
+        this.userAnswer[groupId].forEach((answer) => {
+          if (!includes(this.correct[groupId], answer.id)) {
+            wrongAnswers.push(answer.id);
+            this.answersCollection.push(answer);
+          }
+        });
+        this.removeFromGroup(wrongAnswers, groupId);
```
Comment (abasic) [inline src/teaching-element/Assessment/DragDrop.vue]: If `wrongAnswers` are only used to remove from the `userAnswers` prop then can't the `answersCollection` prop be used instead to remove them?

**Comment by kronicker** [comment]
@kjuej Please, resolve conflicts

File: src/teaching-element/Assessment/DragDrop.vue [hunk 10]
```diff
@@ -99,12 +100,15 @@ export default {
     },
     isBoxFull() {
       return this.answersCollection.length === 0;
+    },
+    removeClass() {
+      return get(this.options.dragDrop, 'removeClass', 'mdi mdi-close');
```
Comment (vladimyr) [inline src/teaching-element/Assessment/DragDrop.vue]: Kind reminder that naming things properly is hard...
There are two properties attached on every DOM element out there
1. [`Element.className`](https://mdn.io/classname)
2. [`Element.classList`](https://mdn.io/classlist)

Judging by your defaults this should be called `removeClassName` instead :bowtie:

**Pull request merged**